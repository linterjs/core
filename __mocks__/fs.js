const realFs = jest.requireActual("fs");
const { vol } = jest.requireActual("memfs");
const { ufs } = jest.requireActual("unionfs");
ufs.use(realFs).use(vol);
module.exports = ufs;
