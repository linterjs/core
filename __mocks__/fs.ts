const fs = require.requireActual("fs");

let files = Object.create(null);

function __clearFiles() {
  files = Object.create(null);
}

function __setFiles(newFiles: object) {
  Object.entries(newFiles).forEach(([path, content]) => {
    files[path] = content;
  });
}

const mockFs = {
  ...fs,
  __clearFiles,
  __setFiles,
  readFileSync: jest.fn((path, options) => {
    if (path in files) {
      return files[path];
    }

    fs.readFileSync(path, options);
  }),
  writeFileSync: jest.fn((path, data, options) => {
    if (path in files) {
      files[path] = data;
    } else {
      fs.writeFileSync(path, data, options);
    }
  })
};

module.exports = mockFs;
