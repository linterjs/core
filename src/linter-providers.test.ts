import { __clearFiles, __setFiles, readFileSync } from "fs";
import {
  LinterAdapter,
  LinterAdapterFormat,
  LinterAdapterLint
} from "./linter-adapter";
import {
  LinterProvider,
  loadLinterProvidersFromFile,
  registerLinterProvider,
  registeredLinterProvidersFilePath,
  deregisterLinterProvider
} from "./linter-providers";

jest.mock("@linter/eslint");
jest.mock("@linter/prettier");
jest.mock("fs");

beforeAll(() => {
  __setFiles({
    [registeredLinterProvidersFilePath]: ""
  });
});

describe("Register and deregister linter providers", () => {
  test("Register @linter/eslint and @linter/prettier", () => {
    registerLinterProvider("@linter/eslint");
    let fileContent = readFileSync(registeredLinterProvidersFilePath, "utf8");
    expect(fileContent).toMatchSnapshot();
    registerLinterProvider("@linter/prettier");
    fileContent = readFileSync(registeredLinterProvidersFilePath, "utf8");
    expect(fileContent).toMatchSnapshot();
  });

  test("Register @linter/eslint again", () => {
    registerLinterProvider("@linter/eslint");
    const fileContent = readFileSync(registeredLinterProvidersFilePath, "utf8");
    expect(fileContent).toMatchSnapshot();
  });

  test("Deregister @linter/eslint and @linter/prettier", () => {
    deregisterLinterProvider("@linter/eslint");
    let fileContent = readFileSync(registeredLinterProvidersFilePath, "utf8");
    expect(fileContent).toMatchSnapshot();
    deregisterLinterProvider("@linter/prettier");
    fileContent = readFileSync(registeredLinterProvidersFilePath, "utf8");
    expect(fileContent).toMatchSnapshot();
  });
});

describe("Load linter providers", () => {
  test("with no registered linter providers file available", () => {
    __clearFiles();

    expect(() => {
      const linterProviders = loadLinterProvidersFromFile();
    }).toThrowErrorMatchingSnapshot();

    __setFiles({
      [registeredLinterProvidersFilePath]: ""
    });
  });

  test("when linter provider module isn't available", () => {
    registerLinterProvider("@linter/non-existent");

    expect(() => {
      const linterProviders = loadLinterProvidersFromFile();
    }).toThrowErrorMatchingSnapshot();

    deregisterLinterProvider("@linter/non-existent");
  });

  test("with @linter/eslint registered", () => {
    registerLinterProvider("@linter/eslint");

    const loadedLinterProviders = loadLinterProvidersFromFile();
    expect(loadedLinterProviders).toMatchSnapshot();

    deregisterLinterProvider("@linter/eslint");
  });
});
