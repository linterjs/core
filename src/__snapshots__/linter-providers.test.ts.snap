// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Load linter providers when linter provider module isn't available 1`] = `"No linters registered. Please install at least one linter provider package."`;

exports[`Load linter providers with @linter/eslint registered 1`] = `
Set {
  Object {
    "linterFactory": [Function],
    "name": "eslint",
  },
}
`;

exports[`Load linter providers with no registered linter providers file available 1`] = `"No linters registered. Please install at least one linter provider package."`;

exports[`Register and deregister linter providers Deregister @linter/eslint and @linter/prettier 1`] = `
"@linter/prettier
"
`;

exports[`Register and deregister linter providers Deregister @linter/eslint and @linter/prettier 2`] = `
"
"
`;

exports[`Register and deregister linter providers Register @linter/eslint again 1`] = `
"@linter/prettier;@linter/eslint
"
`;

exports[`Register and deregister linter providers Register @linter/eslint and @linter/prettier 1`] = `
"@linter/eslint
"
`;

exports[`Register and deregister linter providers Register @linter/eslint and @linter/prettier 2`] = `
"@linter/prettier;@linter/eslint
"
`;
